### ISSUE TITLE
BasicGetResult body memory safety in 6.x+

### ISSUE THREAD

**Issue by Pliner** [rabbitmq/rabbitmq-dotnet-client#994]
Hi,

It looks I've accidentally found an amusing bug. Let me start from this example(by the way, I'm using v6.2.1):

```
using System;
using System.Collections.Generic;
using System.Threading;
using RabbitMQ.Client;


namespace DotNetExperiments
{
    public static class Program
    {
        private static readonly ReadOnlyMemory<byte> SentBody = new byte[] {42, 42, 42, 42, 42};
        
        public static void Main()
        {
            // Boostrap stuff
            var connectionFactory = new ConnectionFactory
            {
                Uri = new Uri("amqps://apwlvbny:ZPrB0qqbBkBE0jOwUAHERRXL39fog2jc@rattlesnake.rmq.cloudamqp.com/apwlvbny")
            };
            var connection = connectionFactory.CreateConnection();
            var model = connection.CreateModel();
            var queue = Guid.NewGuid().ToString("N");
            model.QueueDeclare(queue, false, false);

            // Let's store all received bodies
            var receivedBodies = new List<ReadOnlyMemory<byte>>();
            while (true)
            {
                Thread.Sleep(100);

                // Then every iteration we recheck all received bodies: they should match with sentBody 
                foreach (var receivedBody in receivedBodies) CheckReceivedBody(receivedBody);

                model.BasicPublish("", queue, true, null, SentBody);
                var getResult = model.BasicGet(queue, true);
                if (getResult == null)
                    continue;

                // Store received body
                receivedBodies.Add(getResult.Body);
            }
        }

        private static void CheckReceivedBody(ReadOnlyMemory<byte> receivedBody)
        {
            if (receivedBody.Span.SequenceEqual(SentBody.Span))
                return;
            
            Console.WriteLine($"Mismatch: {string.Join(",", receivedBody.Span.ToArray())} != {string.Join(",", SentBody.Span.ToArray())}");
        }
    }
}
```

After a couple of seconds mismatches have been found:
```
Mismatch: 8,0,0,0,0 != 42,42,42,42,42
Mismatch: 8,0,0,0,0 != 42,42,42,42,42
Mismatch: 8,0,0,0,0 != 42,42,42,42,42
Mismatch: 8,0,0,0,0 != 42,42,42,42,42
Mismatch: 8,0,0,0,0 != 42,42,42,42,42
Mismatch: 8,0,0,0,0 != 42,42,42,42,42
Mismatch: 8,0,0,0,0 != 42,42,42,42,42
Mismatch: 8,0,0,0,0 != 42,42,42,42,42
Mismatch: 8,0,0,0,0 != 42,42,42,42,42
Mismatch: 8,0,0,0,0 != 42,42,42,42,42
Mismatch: 8,0,0,0,0 != 42,42,42,42,42
```

An array under `BasicGetResult.Body` was reused. 

The reason of it might be the following: after a [completion](https://github.com/rabbitmq/rabbitmq-dotnet-client/blob/6.x/projects/RabbitMQ.Client/client/impl/ModelBase.cs#L721) of BasicGet RPC continuation, a command is [disposed](https://github.com/rabbitmq/rabbitmq-dotnet-client/blob/6.x/projects/RabbitMQ.Client/client/impl/Session.cs#L52) and an array under `BasicGetResult.Body` is returned to `ArrayPool`. This makes usage of `BasicGetResult.Body` unpredictable at least(https://github.com/EasyNetQ/EasyNetQ/issues/1153) because it is unsafe to use `BasicGetResult.Body` right after `model.BasicGet` returns it.

It looks like the only fix is to copy a body a [bit before](https://github.com/rabbitmq/rabbitmq-dotnet-client/blob/6.x/projects/RabbitMQ.Client/client/impl/ModelBase.cs#L728) a completion of a continuation.

Could you please confirm the issue?

**Comment by michaelklishin** [comment]
This is a `6.x` client [requirement for regular consumers](https://www.rabbitmq.com/dotnet-api-guide.html#consuming-memory-safety). We could return a copy for `basic.get` responses since `basic.get` by definition means you throw efficiency out the window (there is no way to do polling efficiently).

**Comment by Pliner** [comment]
>This is a 6.x client requirement for regular consumers.

Yep, I'm aware of this and it's fine for consumers which receives message via push from broker.

I would like to stress that it looks like there is no way to safely use `basic.get` response, because the buffer(I mean byte arrays which is used by `ReadOnlyMemory<byte>` in the response) had been is already returned to the pool before we had a chance to read the response.

**Comment by bollhals** [comment]
For 6.x it's a valid issue from what I can see which got fixed by #919 which references #918.

**Comment by Pliner** [comment]
@bollhals Have #919 had a chance to be backported to 6.x? Or #995  is preferred way to fix?

**Comment by bollhals** [comment]
We'd have to investigate how much changed already between 6.x and #919, in general though, it should be possible. One thing though is that `BasicGetResult` would have to add IDisposable, which is a breaking change, so if that is approved, I think this should be doable.
(Your fix for sure is a quick fix that does the trick as well)

**Comment by michaelklishin** [comment]
So do we want to accept #995 (thank you for taking the time to contribute, @Pliner!)? It seems reasonable to me because polling and efficiency really do not go hand in hand :) @stebet @bollhals @bording

**Comment by micdenny** [comment]
I would accept #995 me too, because it's easier, and wait the better fix on 8.x

**Comment by bollhals** [comment]
Fine by me 👍🏼

**Comment by TheMulti0** [comment]
Hi, I'm encountering this issue too. 
Is there any information known about when will the fix be released on Nuget?
Thanks!

**Comment by TheMulti0** [comment]
Apparently my issue was different - it had been passing the body between different threads which resulted in the data getting garbaged.
Thanks!

**Comment by Pliner** [comment]
The [fix](https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/995) was released as part of 6.2.2.
